### PR TITLE
Add support for builtin libcurl and some other fixes

### PR DESCRIPTION
**Pull request by Cloudef**
* Fix crash on easy.do() which does not pass the proper type to curl
* Fix compile on latest zig
* Use standard optimize options in built.zig
* Fix typo in errors.zig
* Allow the use of certificates in std.crypto.Certificate.Bundle instead of default one
* Include builtin libcurl compiled against mbedTls, the use of std.crypto.Certificate.Bundle is required, or CURLOPT_CAINFO should be used to point to cacert bundle.

### CODE REVIEW

**Comment by Cloudef** [comment]
Note that the zigcurl dependency points to my fork and should be updated once https://github.com/star-tek-mb/zigcurl/pull/2 is merged.

File: build.zig.zon [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    .version = "0.0.1",
+    .paths = .{
+        "src",
+        "build.zig",
+        "build.zig.zon",
+        "LICENSE",
+    },
+    .dependencies = .{
+        .libcurl = .{
+            .url = "https://github.com/Cloudef/zigcurl/archive/e9e726131d3eb80c02d20b7ee5ca6935324d6697.tar.gz",
```
Comment (jiacai2050) [inline build.zig.zon:12]: Which libcurl version does this bundle with?

File: build.zig.zon [hunk 2]
```diff
@@ -0,0 +1,16 @@
+.{
+    .name = "zig-curl",
+    .version = "0.0.1",
```
Comment (jiacai2050) [inline build.zig.zon:3]: 0.1.0

File: build.zig.zon (hunk 1 above)
Comment (Cloudef) [reply build.zig.zon:12]: curl 8.1.1
mbedtls 3.4.0
zlib 1.2.13
https://github.com/star-tek-mb/zigcurl

File: build.zig [hunk 3]
```diff
@@ -7,20 +7,24 @@ const MODULE_NAME = "curl";
 
 pub fn build(b: *std.Build) void {
     const target = b.standardTargetOptions(.{});
+    const optimize = b.standardOptimizeOption(.{});
```
Comment (jiacai2050) [inline build.zig:10]: Does this affect how libcurl is compile?

Comment (Cloudef) [reply build.zig:10]: I don't know about that, but I kept getting warnings from zig compiler when using zig-curl as dependency and passing my own optimize flags unless I did this.

File: src/main.zig [hunk 4]
```diff
@@ -6,6 +6,38 @@ pub const Easy = @import("easy.zig");
 pub usingnamespace Easy;
 pub usingnamespace @import("c.zig");
 
+pub var CaBundle: ?[]const u8 = null;
```
Comment (jiacai2050) [inline src/main.zig:9]: Could we avoid this global variable?

I would like to init ca bundle when init `Easy` instance, just like `timeout_ms`, `default_user_agent`.

File: src/main.zig [hunk 5]
```diff
@@ -6,6 +6,38 @@ pub const Easy = @import("easy.zig");
...
+
+/// Not thread safe
+/// Refreshes CaBundle for future requests
+pub fn refresh_ca_bundle() !void {
+    if (CaBundle) |ptr| std.heap.c_allocator.free(ptr);
+    var bundle = std.crypto.Certificate.Bundle{};
+    defer bundle.deinit(std.heap.c_allocator);
+    try bundle.rescan(std.heap.c_allocator);
```
Comment (jiacai2050) [inline src/main.zig:17]: It seems `rescan` just scan certs on host machine, what's the difference when we call libcurl directly?

File: build.zig.zon (hunk 1 above)
Comment (jiacai2050) [reply build.zig.zon:12]: Maybe we need to add this in README.

File: src/main.zig (hunk 5 above)
Comment (Cloudef) [reply src/main.zig:17]: The curl isn't built with cacert path so we need to pass it certs in order for TLS verification to work. Using zig's std.crypto.Certificate.bundle is portable way of doing this. You can also point it to your system's cacert with curlopt if you want but the application has to figure out where that is for each OS.

Comment (Cloudef) [reply src/main.zig:17]: Looking at libcurl's source code, it seems the CURL_CA_FALLBACK only works for openssl backend, and for non linux platforms the defaults come from system's SSL library(?). The built system can autodetect cabundle from unix paths but that's about it.

Comment (Cloudef) [reply src/main.zig:17]: I was also thinking of verifying certs manually using the std.crypto.Certificate.bundle api, but I was not familiar enough with libcurl to do this, so the simplest way was just to re-encode the certs zig already decodes :smile:

File: src/main.zig (hunk 4 above)
Comment (Cloudef) [reply src/main.zig:9]: That might be the better way to go. Then you can simply create new Easy instance if you want refreshed ca bundle.

File: build.zig.zon (hunk 1 above)
Comment (Cloudef) [reply build.zig.zon:12]: Another option could be to merge the work from that repo to this, that way you wouldn't have to depend on anything. I'm pretty sure the author's fine with it.

**Comment by Cloudef** [comment]
Side-note, it would be nice to have std.http.Client's fetch compatible api. I use this in my own project until std.crypto has TLSv1.2 support:
```zig
// Temporary crap until std.http.Client has TLSv1.2 support
// https://github.com/ziglang/zig/issues/14172#issuecomment-1384567705
const curl = @import("curl");
const Client = struct {
    client: curl.Easy,

    pub fn init(allocator: std.mem.Allocator) !@This() {
        return .{ .client = try curl.Easy.init(allocator) };
    }

    pub fn deinit(self: *@This()) void {
        self.client.deinit();
    }

    pub fn fetch(self: *@This(), allocator: std.mem.Allocator, options: std.http.Client.FetchOptions) !std.http.Client.FetchResult {
        var reader = switch (options.payload) {
            .file => |file| file.reader().any(),
            .string => |str| blk: {
                var stream = std.io.fixedBufferStream(str);
                break :blk stream.reader().any();
            },
            .none => blk: {
                const empty: []const u8 = &.{};
                var stream = std.io.fixedBufferStream(empty);
                break :blk stream.reader().any();
            },
        };

        const target = switch (options.location) {
            .url => |url| url,
            .uri => @panic("unsupported"),
        };

        var headers = curl.Easy.RequestHeader.init(allocator);
        for (options.headers.list.items[0..]) |header| try headers.add(header.name, header.value);
        var req = curl.Easy.Request(*@TypeOf(reader)).init(target, &reader, .{
            .method = std.meta.stringToEnum(curl.Easy.Method, @tagName(options.method)).?,
            .header = headers,
        });
        defer req.deinit();

        const res = try self.client.do(req);
        return .{
            .allocator = allocator,
            .options = options,
            .status = @enumFromInt(res.status_code),
            .body = res.body.items,
            .headers = .{ .allocator = allocator },
        };
    }
};
```

**Comment by jiacai2050** [comment]
> it would be nice to have std.http.Client's fetch compatible api.

This is a nice point, we can do this in another PR.

File: src/main.zig (hunk 5 above)
Comment (Cloudef) [reply src/main.zig:17]: I wonder if we should check if cainfo or capath is set in curl_version_info and if so skip using std.crypto, or should I make the Easy init() method have a options struct with field whether to use std.crypto, defaulting to true https://curl.se/libcurl/c/curl_version_info.html

File: src/main.zig (hunk 4 above)
Comment (jiacai2050) [reply src/main.zig:9]: @Cloudef Hi, do you still have time to push this further?

I can help if you are busying.

Comment (Cloudef) [reply src/main.zig:9]: Yes I have time today. From tomorrow maybe not so much :slightly_smiling_face:

File: src/main.zig (hunk 5 above)
Comment (jiacai2050) [reply src/main.zig:17]: Sound great, I think you could implement this in next PR.

For this PR, only add libcurl, no more.

**Comment by Cloudef** [comment]
The `std.crypto.Certificate.Bundle` usage is now under EasyOptions setting. I think this is the best way instead of trying to auto-detect what the library user might want. It makes sense to default to true here though as the included curl won't be very useful without disabling TLS verification with curlopt otherwise.
https://github.com/jiacai2050/zig-curl/blob/96ba813d61d0817fa755df5d299befb08b4b0fba/src/easy.zig#L233-L263

**Review by jiacai2050 — APPROVED** [review]
LGTM

**Comment by jiacai2050** [comment]
I will do some polish based on your branch, please don't update any more.

**Review by jiacai2050 — APPROVED** [review]

**Comment by Cloudef** [comment]
I think it's possible. In the meantime https://github.com/star-tek-mb/zigcurl/pull/2 the PR with fixes was merged in.

**Comment by jiacai2050** [comment]
@Cloudef Hi, just let you know, I already vendor those C libraries. 

I choose not to use submodule since those libraries have really long commit history, and use a [update.sh](https://github.com/jiacai2050/zig-curl/blob/main/libs/update.sh) script to deal with the update.

**Pull request merged**